server/server.go

```diff
@@ -682,58 +675,55 @@ func (s *Server) Serve(l net.Listener) error {
 		}
 	}
 
-	// This map keeps track of all existing connections. This allows us to close
-	// them all (and thus cause the associated goroutines to quit, freeing
-	// resources) when we're about to return. The mutex protects the map since the
-	// handle method concurrently deletes its connection from it once the
-	// connection dies.
-	var mapMtx sync.Mutex
-	conns := make(map[*client.ConnHandle]bool)
-	var wg sync.WaitGroup
-
-	defer func() {
-		// Close all of the connections so that the associated goroutines quit.
-		mapMtx.Lock()
-		for c := range conns {
-			c.Destroy()
-		}
-		mapMtx.Unlock()
-		// Wait for all of the goroutines to quit.
-		wg.Wait()
-	}()
+	// Perform the TLS handshake explicitly so we can determine if this is a
+	// limited connection.
+	tconn := tls.Server(c, s.tlsConfig)
+	err := tconn.Handshake()
+	if err != nil {
+		log.Errorf("connection %v: handshake failed: %v", c.RemoteAddr(), err)
+		tconn.Close()
+		return
+	}
+	limited, err := s.config.isLimited(tconn.ConnectionState())
+	if err != nil {
+		log.Errorf("connection %v: could not determine if limited: %v", c.RemoteAddr(), err)
+		tconn.Close()
+		return
+	}
 
-	for {
-		c, err := accept(l)
-		if err != nil {
-			log.Errorf("Accept error: %v; shutting down server", err)
-			return err
-		}
-		wg.Add(1)
-		go func() {
-			tconn := tls.Server(c, s.tlsConfig) // If limited use just limited workers
-			var conn *conn
-			if s.config.isLimited(tconn) {
-				log.Debug("Connection is limited")
-				conn = newConn(s, c.RemoteAddr().String(), tconn, timeout, s.wp.Limited, s.wp.Limited)
-			} else {
-				conn = newConn(s, c.RemoteAddr().String(), tconn, timeout, s.wp.ECDSA, s.wp.Other)
-			}
+	var connStr string
+	var conn *conn
+	if limited {
+		connStr = fmt.Sprintf("limited connection %v", c.RemoteAddr())
+		conn = newConn(s, c.RemoteAddr().String(), tconn, timeout, s.wp.Limited, s.wp.Limited)
+	} else {
+		connStr = fmt.Sprintf("connection %v", c.RemoteAddr())
+		conn = newConn(s, c.RemoteAddr().String(), tconn, timeout, s.wp.ECDSA, s.wp.Other)
+	}
 
-			log.Debugf("connection %v: spawned", c.RemoteAddr())
-			handle := client.SpawnConn(conn)
+	// Acquire the lock to atomically spawn the reader/writer goroutines for
+	// this connenction and add it to the connections map.
+	s.mtx.Lock()
+	if s.shutdown {
+		s.mtx.Unlock()
+		log.Debugf("%s: rejected (server is shutting down)", connStr)
+		tconn.Close()
+		return
+	}
+	handle := client.SpawnConn(conn)
+	s.listeners[l][handle] = struct{}{}
+	s.mtx.Unlock()
+	log.Debugf("%s: spawned", connStr)
 
-			mapMtx.Lock()
-			conns[handle] = true
-			mapMtx.Unlock()
+	// Block here until the connection and associated goroutines have completed.
+	handle.Wait()
+	log.Debugf("%s: removed", connStr)
```

Comment:
Hang on, we're logging this line before we actually removed the connection. Maybe we should reverse the order to help in debugging.

Reply:
Are you sure? Once `Wait` returns it means both the reader and writer goroutines have exited.

Reply:
We killed it, but we didn't take it out of the structure yet.

Reply:
Oh that's what you mean by removed. I generally care more about when the connection is closed rather than actually removed from the map, but I can see value in either. I'll log both steps.